sdk/python/tests/README.md

```diff
@@ -0,0 +1,341 @@
+
+1. how Feast tests are setup
+2. how to extend the test suite to test new functionality
+3. how to use the existing test suite to test a new custom offline / online store.
+
+## Test suite overview
+
+Let's inspect the test setup in `sdk/python/tests/integration`:
+
+```bash
+$ tree
+.
+├── e2e
+│   ├── test_go_feature_server.py
+│   ├── test_python_feature_server.py
+│   ├── test_universal_e2e.py
+│   ├── test_usage_e2e.py
+│   └── test_validation.py
+├── feature_repos
+│   ├── integration_test_repo_config.py
+│   ├── repo_configuration.py
+│   └── universal
+│       ├── catalog
+│       ├── data_source_creator.py
+│       ├── data_sources
+│       │   ├── __init__.py
+│       │   ├── bigquery.py
+│       │   ├── file.py
+│       │   ├── redshift.py
+│       │   └── snowflake.py
+│       ├── entities.py
+│       ├── feature_views.py
+│       ├── online_store
+│       │   ├── __init__.py
+│       │   ├── datastore.py
+│       │   ├── dynamodb.py
+│       │   ├── hbase.py
+│       │   └── redis.py
+│       └── online_store_creator.py
+├── materialization
+│   └── test_lambda.py
+├── offline_store
+│   ├── test_feature_logging.py
+│   ├── test_offline_write.py
+│   ├── test_push_features_to_offline_store.py
+│   ├── test_s3_custom_endpoint.py
+│   └── test_universal_historical_retrieval.py
+├── online_store
+│   ├── test_online_retrieval.py
+│   ├── test_push_features_to_online_store.py
+│   └── test_universal_online.py
+└── registration
+    ├── test_feature_store.py
+    ├── test_inference.py
+    ├── test_registry.py
+    ├── test_sql_registry.py
+    ├── test_universal_cli.py
+    ├── test_universal_odfv_feature_inference.py
+    └── test_universal_types.py
+
+```
+
+- `feature_repos` has setup files for most tests in the test suite.
+- `conftest.py` and some of the individual test files contain fixtures which can be used to  on different offline stores, online stores, etc. and thus abstract away store specific implementations so we don't need to rewrite the same test implementation for different stores.
+
+## Structure of the test suite
+
+### What is the universal test suite?
+
+The universal test suite verifies that crucial Feast functions(e.g `get_historical_features`, `get_online_features` etc.) have the correct behavior for each of the different environments that Feast could be used in. These environments are combinations of an offline store, online store, and provider and the universal test suite serves to run basic functional verification against all of these different permutations.
+
+We use pytest [fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) to accomplish this without writing excess code.
+
+Tests in Feast are split into integration and unit tests.
+
+### Is it an integration or unit test?
+
+- In an integration test, multiple modules of Feast are tested together. Integration tests mainly involve testing of Feast components that connect to services outside of Feast(e.g connecting to gcp or aws clients).
```

Comment:
nit: prefer full sentences in paragraph as opposed to bullet points here (to be consistent with what we do elsewhere)

Reply:
also this is somewhat misleading - I think we decided that integration = non-local / requires external services, unit = local

specifically, "In an integration test, multiple modules of Feast are tested together." is misleading since it implies that `unit/local_feast_tests/test_e2e_local.py` would be an integration test instead of unit

Reply:
actually scratch the paragraph thing; bullet points seem fine!

Reply:
fixed